core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -348,30 +348,49 @@ public static boolean similar(String s, String pattern, String escape) {
 
   // =
 
+  /** SQL = operator applied to BigDecimal values (neither may be null). */
+  public static boolean eq(BigDecimal b0, BigDecimal b1) {
+    return b0.stripTrailingZeros().equals(b1.stripTrailingZeros());
+  }
+
   /** SQL = operator applied to Object values (including String; neither
    * side may be null). */
   public static boolean eq(Object b0, Object b1) {
-    return b0.equals(b1);
+    if (b0.getClass().equals(b1.getClass())) {
+      // The result of SqlFunctions.eq(BigDecimal, BigDecimal) is making more sense than
+      // BigDecimal.equals(BigDecimal).
+      // So if both of types are BigDecimal, we just use SqlFunctions.eq(BigDecimal, BigDecimal).
+      if (BigDecimal.class.isAssignableFrom(b0.getClass())) {
```

Comment:
In this and a few other places you could use `aClass.isInstance(b)` rather than `aClass.isAssignableFrom(b.getClass())`. The two are equivalent, but the first is shorter and possibly quicker.